samza-sql-shell/src/main/java/org/apache/samza/sql/client/interfaces/EnvironmentVariableHandlerImpl.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.sql.client.interfaces;
+
+import org.apache.samza.sql.client.util.CliException;
+import org.apache.samza.sql.client.util.Pair;
+
+import java.util.*;
+
+public abstract class EnvironmentVariableHandlerImpl implements EnvironmentVariableHandler{
+  protected EnvironmentVariableSpecs specs = new EnvironmentVariableSpecs();
+  protected Map<String, String> envVars = new HashMap<>();
+
+  @Override
+  public int setEnvironmentVariable(String envName, String value) {
+    EnvironmentVariableSpecs.Spec spec = specs.getSpec(envName);
+    if(spec == null) {
+      if(isAcceptUnknowName()) {
+        return setEnvironmentVariableHelper(envName, value);
+      } else {
+        return -1;
+      }
+    }
+
+    String[] possibleValues = spec.getPossibleValues();
+    if (possibleValues == null || possibleValues.length == 0) {
+      return setEnvironmentVariableHelper(envName, value);
+    }
+
+    for (String s : possibleValues) {
+      if (s.equalsIgnoreCase(value)) {
+        if(!processEnvironmentVariable(envName, value)) {
+          throw new CliException(); // should not reach here
+        }
+        envVars.put(envName, value);
+        return 0;
+      }
+    }
+    return -2;
+  }
+
+  @Override
+  public String getEnvironmentVariable(String envName) {
+    return envVars.get(envName);
+  }
+
+  @Override
+  public List<Pair<String, String>> getAllEnvironmentVariables() {
+    List<Pair<String, String>> list = new ArrayList<>();
+    Iterator<Map.Entry<String, String>> it =  envVars.entrySet().iterator();
+    while(it.hasNext()) {
+      Map.Entry<String, String> entry = it.next();
+      list.add(new Pair<>(entry.getKey(), entry.getValue()));
+    }
+    return list;
+  }
+
+  @Override
+  public EnvironmentVariableSpecs getSpecs() {
+    return specs;
+  }
+
+  /**
+   * Subclasses shall override this method to process their settings
+   * @param envName name of the environment variable
+   * @param value value of the environment variable
+   * @return succeed or not
+   */
+  protected boolean processEnvironmentVariable(String envName, String value) {
```

Comment:
Can we use exception handling for success or failure that ways it's more consistent with other parts of the code?

Reply:
At the moment Samza-Sql-Shell prefers return values for expected error or must-handle errors. I know it's controversial but as I mentioned in a previous comment, we need to change the way of reporting errors entirely if we decide to use exceptions. Can we keep this issue for a while and see what we can do as a future work?

Reply:
I think our offline discussion should clarify this as well :).